### PR TITLE
ENH: HaarPSI FOM

### PR DESCRIPTION
**Pull request by kohr-h**
Implementation of the HaarPSI figure of merit, also gave me an opportunity to play around with pytorch a bit. Seems to work quite well, one can now use HaarPSI as output layer and (optionally) train its parameters, whether or not that makes sense.

**TODO**
- [x] Add proper example for using HaarPSI
- [x] ~~Make sure the convolution in pytorch does what it should~~ Moved out
- [x] In `contrib/fom`: write `S` function with `c**2` instead of `c`, makes it more intuitive.
- [x] In `contrib/fom`: rename some functions to be HaarPSI-specific
- [x] Use `scipy` functions where applicable
- [x] Simplify kernel FFT manipulation

### CODE REVIEW

**Comment by kohr-h** [comment]
Ready for review, but as contrib stuff this could actually go straight in.

File: odl/contrib/fom/util.py [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+    """The inverse of the logistic function.
+
+    It is defined as ::
+
+        l_a^{-1}(x) = (log(x) - log(1 - x)) / a
+    """
+    return (np.log(x) - np.log(1 - x)) / a
+
+
+def similarity_map(img1, img2, axis, c, a):
```
Comment (adler-j) [inline odl/contrib/fom/util.py]: This name is way to general.

Frankly if it is only used in haarpsi, it should be internal and/or private

File: odl/contrib/fom/util.py [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+    """Local similarity score for directional features along an axis.
+
+    Parameters
+    ----------
+    img1, img2 : array-like
+        The images to compare. They must have equal shape.
+    axis : {0, 1}
```
Comment (adler-j) [inline odl/contrib/fom/util.py]: Any reason this is 2d only?

File: odl/contrib/fom/util.py [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+    # Compute similarity scores for the levels
+    sim_lvl1 = similarity_score(np.abs(img1_lvl1), np.abs(img2_lvl1), c)
+    sim_lvl2 = similarity_score(np.abs(img1_lvl2), np.abs(img2_lvl2), c)
+
+    # Return logistic of the mean value
+    sim = (sim_lvl1 + sim_lvl2) / 2
+    return logistic(sim, a)
+
+
+def similarity_weight_map(img1, img2, axis):
```
Comment (adler-j) [inline odl/contrib/fom/util.py]: See above comment. These are haarpsi specific.

File: odl/contrib/pytorch/functions.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
+# Copyright 2014-2017 The ODL contributors
+#
+# This file is part of ODL.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+"""Predefined functions implementing custom automatic differentiation.
```
Comment (adler-j) [inline odl/contrib/pytorch/functions.py]: Empty file?

File: odl/contrib/pytorch/modules/basic.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+"""Basic pytorch modules for general-purpose use."""
+
+import torch
+from torch import nn
+
+__all__ = ('Logistic', 'InvLogistic')
+
+
+class Logistic(nn.Module):
```
Comment (adler-j) [inline odl/contrib/pytorch/modules/basic.py]: Not at all sure if ODL should have stuff like this that we expose "externally" so to say. If we need it internally, it should be in some util folder.

We don't want to end up as a general purpose pytorch library (i guess?)

For TF I'm hosting my stuff at
https://github.com/adler-j/adler

File: odl/contrib/pytorch/modules/haarpsi.py [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+        conv_l3 = nn.functional.conv2d(xy, f_l3, padding=7)
+
+        conv_x_l3 = conv_l3[0, 0, 3:-4, 3:-4]
+        conv_y_l3 = conv_l3[1, 0, 3:-4, 3:-4]
+
+        # Return maximum of the absolute values
+        return torch.max(torch.abs(conv_x_l3), torch.abs(conv_y_l3))
+
+
+class HaarPSI(nn.Module):
```
Comment (adler-j) [inline odl/contrib/pytorch/modules/haarpsi.py]: While I really like this feature. I'm not sure if we should have it in ODL.

One option would frankly be if we (aka, @kohr-h, @adler-j and whoever wants in) created a new library specifically for stuff like this that we can share. One option would be to "steal" the stuff from the adler library as a "base" for the tensorflow stuff and then add stuff like this for torch. By import "gating" we could easily have it all in one library.

File: odl/contrib/fom/examples/haarpsi.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
+"""Demonstration of the components of the HaarPSI figure of merit."""
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/contrib/fom/examples/haarpsi.py]: Apart from having ~100 lines, there is nothing complex about this example. It just evaluates a bunch of functions and plots the results. And I find it actually kind of instructive to look at the comparison images themselves.

File: odl/contrib/fom/examples/supervised_comparison.py [hunk 8]
```diff
@@ -36,53 +38,52 @@
...
-plt.xlabel('Noise level')
-plt.ylabel('FOM')
+    ssim.append(
+        fom.ssim(phantom_noisy, phantom, normalized=True))
+
+    psnr.append(
+        fom.psnr(phantom_noisy, phantom, normalize=True))
+
+    haarpsi.append(
+        fom.haarpsi(phantom_noisy, phantom, alpha=4))
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/contrib/fom/examples/supervised_comparison.py]: Dammit, should be `a`. Can actually be just left out.

File: odl/contrib/fom/examples/supervised_comparison.py [hunk 9]
```diff
@@ -36,53 +38,52 @@
...
+
+fig, ax = plt.subplots()
+ax.plot(mse)
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/contrib/fom/examples/supervised_comparison.py]: I'll try, but I don't know if the labels actually appear then.

File: odl/contrib/fom/util.py [hunk 10]
```diff
@@ -0,0 +1,437 @@
...
+
+"""Utility functions for FOMs."""
+
+import numpy as np
+
+try:
+    import pyfftw
+except ImportError:
+    pyfftw = None
+    PYFFTW_AVAILABLE = False
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/contrib/fom/util.py]: Yeah, leftover

File: odl/contrib/fom/util.py [hunk 11]
```diff
@@ -0,0 +1,437 @@
...
+    else:
+        raise ValueError('unsupported `impl` {!r}'.format(impl_in))
+
+    if padding:
+        return conv[padding:-padding, padding:-padding]
+    else:
+        return conv
+
+
+def similarity_score(x, y, c):
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/contrib/fom/util.py]: That was basically for lack of a better name. I can also just inline the function.

File: odl/contrib/fom/util.py [hunk 12]
```diff
@@ -0,0 +1,437 @@
...
+    *A Haar Wavelet-Based Perceptual Similarity Index for Image Quality
+    Assessment*. arXiv:1607.06140 [cs], Jul. 2016.
+    """
+    x = np.asarray(x)
+    y = np.asarray(y)
+    c = float(c)
+    return (2 * x * y + c ** 2) / (x ** 2 + y ** 2 + c ** 2)
+
+
+def logistic(x, a):
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/contrib/fom/util.py]: Scipy function 👍

File: odl/contrib/fom/util.py [hunk 13]
```diff
@@ -0,0 +1,437 @@
...
+    """The logistic function.
+
+    It is defined as ::
+
+        l_a(x) = 1 / (1 + exp(-a * x))
+    """
+    return 1 / (1 + np.exp(-a * x))
+
+
+def inv_logistic(x, a):
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/contrib/fom/util.py]: Amazing what scipy has 😉

File: odl/contrib/pytorch/functions.py (hunk 4 above)
Comment (kohr-h) [reply odl/contrib/pytorch/functions.py]: Well I moved around code until I fully understood what `Function` and `Module` objects are there for. But then I had compiled some explaining text that I found helpful.
That said, I could just add the documentation to the `operator.py` file, or the other way around, add `TorchOperator` to `functions.py`.

File: odl/contrib/pytorch/modules/haarpsi.py (hunk 6 above)
Comment (kohr-h) [reply odl/contrib/pytorch/modules/haarpsi.py]: Sounds good, a deep learning addon library. I don't mind not including the stuff in ODL. I'm also all for hosting something together, since we can do reviews then 😛

File: odl/contrib/fom/examples/supervised_comparison.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (adler-j) [reply odl/contrib/fom/examples/supervised_comparison.py]: Need to call `plt.legend()` in the end

**Comment by kohr-h** [comment]
I've moved out all `pytorch` stuff to [a new package](https://github.com/odlgroup/deep-learning-library), and the renaming `layer.py -> operator.py` to a separate PR #1131 
Will look after the comments now.

File: odl/contrib/fom/test/test_supervised.py [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+                   odl.util.testutils.skip_if_no_pyfftw('pyfftw')]
+fft_impl = simple_fixture('fft_impl', fft_impl_params,
+                          fmt=" {name} = '{value.args[1]}' ")
+
+
+def filter_image(image, fh, fv):
+    """Reference filtering function using ``scipy.signal.convolve``."""
+    fh, fv = np.asarray(fh), np.asarray(fv)
+    image = scipy.signal.convolve(image, fh[:, None], mode='same')
+    return scipy.signal.convolve(image, fv[None, :], mode='same')
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/contrib/fom/test/test_supervised.py]: I'm not gonna change this. First, `sepfir2d` uses a different boundary condition, and second, this is only for comparison.

File: odl/contrib/fom/test/test_supervised.py [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+def test_filter_image_fft(fft_impl):
+    """Test the FFT filtering function against the real-space variant."""
+    image = np.random.rand(128, 128)
+    image -= np.mean(image)
+
+    for fh, fv in zip([[1, 1], [1, 1, 1], [1, 1, 1, 1]],
+                      [[-1, 1], [-1, 1, 1], [-1, -1, 1, 1]]):
+
+        conv_real = filter_image(image, fh, fv)
+        conv_fft = filter_image_fft(image, fh, fv, impl=fft_impl, padding=16)
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/contrib/fom/test/test_supervised.py]: I'll probably go with `min(max(len(fh), len(fv)) - 1, 64)` or so and default to `None`. Usually, `len(ker) - 1` padding is enough to get equality up to 15 digits or so.

File: odl/contrib/fom/util.py (hunk 1 above)
Comment (kohr-h) [reply odl/contrib/fom/util.py]: Okay with `haarpsi_similarity_map` or do you think it should be `_haarpsi_similarity_map`? I can also just not expose the functions in `__all__`.

File: odl/contrib/fom/util.py (hunk 2 above)
Comment (kohr-h) [reply odl/contrib/fom/util.py]: Well I didn't bother to do nD for now. Wouldn't be a big problem. For now I added TODO comments.

**Comment by kohr-h** [comment]
Everything addressed. Ready for re-review (or merge).

File: odl/contrib/fom/supervised.py [hunk 16]
```diff
@@ -636,7 +634,88 @@ def psnr(data, ground_truth, normalize=False):
     else:
         return 20 * np.log10(max_true) - 10 * np.log10(mse_result)
 
+
+def haarpsi(data, ground_truth, a=4.0, c=None):
```
Comment (kohr-h) [inline odl/contrib/fom/supervised.py]: Maybe we can go with the default `4.2` as in the article.

Comment (adler-j) [reply odl/contrib/fom/supervised.py]: We should indeed follow the article.

**Comment by kohr-h** [comment]
Done, merging after CI
